### ISSUE TITLE
2.7.0-beta3 cannot connect if E2EE-app 1.5.2-beta1 is activated

### ISSUE THREAD

**Issue by spackmat** [nextcloud/desktop#2223]
Hi, I am testing the new E2EE with Windows client 2.7.0-beta3 and E2EE Server-App 1.5.2-beta1 on NextCloud Release 19.0.1. 

I removed the old E2EE-App from the server and cleaned up as stated in https://help.nextcloud.com/t/help-test-the-latest-version-of-e2ee/87590, also removed the whole old windows desktop client and found and removed two nextcloud related entries in the windows login information store (don't know the English name for it, in German it is Anmeldeinformationsverwaltung).

Then I installed and activated the 1.5.2-beta1 of the E2EE app on the server and the 2.7.0-beta3 of the desktop app. This opened my browser and I added allowed the access.

### Expected behaviour
I expected the folder to sync.

### Actual behaviour
Instead the sync doesn't start. My account is registered inside the desktop app and on the add account screens it also told me the correct count of files on the server for that account. But the icon is grey and on the settings screen it remains in connecting state.

When I deactivate the E2EE app on the server and restart the client, it starts to synchronize and everything works well. When I enable the E2EE app again, the connection cannot be established again. In the logs I can see that a new keypair is created but cannot be uploaded with a 409 response.

### Steps to reproduce
1. Have NextCloud 19.0.1 with E2EE app 1.5.2-beta1 activated and the windows client 2.7.0-beta3 running.
2. Deactivate and activate the E2EE on the server and observe the client connecting or not.

### Client configuration
Client version: 2.7.0-beta3

Operating system: Windows 10 Pro 1910

OS language: German

Qt version used by client package (Linux only, see also Settings dialog):

Client package (From Nextcloud or distro) (Linux only):

Installation path of client: C:\Program Files\NextCloud

### Server configuration
Nextcloud version: 19.0.1

Storage backend (external storage):

### Logs

Please use Gist (https://gist.github.com/) or a similar code paster for longer
logs.

1. Client logfile:

````
2020-07-21 16:26:24:579 [ info nextcloud.gui.folder.manager ]:	Setup folders from settings file
2020-07-21 16:26:24:580 [ info nextcloud.sync.database ]:	Detected filesystem "NTFS" for "C:/Users/myuser/AppData/Roaming/Nextcloud/._sync_1148a9071c01.db"
2020-07-21 16:26:24:581 [ info nextcloud.sync.configfile ]:	Adding user defined ignore list to csync: "C:/Users/myuser/AppData/Roaming/Nextcloud/sync-exclude.lst"
2020-07-21 16:26:24:583 [ info nextcloud.gui.folder.manager ]:	Adding folder to Folder Map  OCC::Folder(0x28a6d1a6650) "1"
2020-07-21 16:26:24:583 [ info nextcloud.gui.folder.manager ]:	Schedule folder  "1"  to sync!
2020-07-21 16:26:24:583 [ info nextcloud.gui.folder.manager ]:	Folder is not ready to sync, not scheduled!
2020-07-21 16:26:24:586 [ info nextcloud.gui.application ]:	Sync state changed for folder  "https://nextcloud.myserver.lan/remote.php/webdav/" :  "Not yet Started"
2020-07-21 16:26:24:588 [ info nextcloud.sync.database ]:	sqlite3 version "3.24.0"
2020-07-21 16:26:24:588 [ info nextcloud.sync.database ]:	sqlite3 journal_mode= "wal"
2020-07-21 16:26:24:589 [ info nextcloud.sync.clientproxy ]:	Set proxy configuration to use the preferred system proxy for http tcp connections
2020-07-21 16:26:24:658 [ warning default ]:	qrc:/qml/src/gui/tray/UserLine.qml:46: TypeError: Cannot read property 'width' of undefined
2020-07-21 16:26:24:674 [ info nextcloud.sync.credentials.webflow ]:	Fetch from keychain!
2020-07-21 16:26:24:683 [ info nextcloud.gui.socketapi ]:	New connection QLocalSocket(0x28a6d504a10)
2020-07-21 16:26:24:684 [ warning nextcloud.sync.credentials.keychainchunk ]:	Unable to read "Nextcloud_myaccount_clientCertificatePEM:https://nextcloud.myserver.lan/:0" chunk "0" "Password entry not found"
2020-07-21 16:26:24:688 [ info nextcloud.gui.folder.navigationpane ]:	Explorer Cloud storage provider: saving path "C:\\Users\\myuser\\Nextcloud" to CLSID "{666aa759-8c2d-48bf-bc8a-8580e620e3de}"
2020-07-21 16:26:24:689 [ warning nextcloud.sync.credentials.keychainchunk ]:	Unable to read "Nextcloud_myaccount_clientKeyPEM:https://nextcloud.myserver.lan/:0" chunk "0" "Password entry not found"
2020-07-21 16:26:24:689 [ warning nextcloud.sync.credentials.webflow ]:	Unable to read client key "Password entry not found"
2020-07-21 16:26:24:692 [ warning nextcloud.sync.credentials.keychainchunk ]:	Unable to read "Nextcloud_myaccount_clientCaCertificatePEM0:https://nextcloud.myserver.lan/:0" chunk "0" "Password entry not found"
2020-07-21 16:26:24:692 [ warning nextcloud.sync.credentials.webflow ]:	Unable to read client CA cert slot "0" "Password entry not found"
2020-07-21 16:26:24:693 [ info nextcloud.gui.account.state ]:	Fetched credentials for "https://nextcloud.myserver.lan" attempting to connect
2020-07-21 16:26:24:693 [ info nextcloud.sync.credentials.webflow ]:	Get QNAM
2020-07-21 16:26:24:696 [ info nextcloud.sync.connectionvalidator ]:	No system proxy set by OS
2020-07-21 16:26:24:696 [ info nextcloud.sync.accessmanager ]:	2 "" "https://nextcloud.myserver.lan/status.php" has X-Request-ID "10927fd5-59ec-4072-acc5-8ec190f1bad7"
2020-07-21 16:26:24:697 [ info nextcloud.sync.networkjob ]:	OCC::CheckServerJob created for "https://nextcloud.myserver.lan" + "status.php" "OCC::ConnectionValidator"
2020-07-21 16:26:24:790 [ info nextcloud.sync.credentials.webflow ]:	request finished
2020-07-21 16:26:24:790 [ info nextcloud.sync.networkjob.checkserver ]:	status.php returns:  QJsonDocument({"edition":"","extendedSupport":false,"installed":true,"maintenance":false,"needsDbUpgrade":false,"productname":"Nextcloud","version":"19.0.1.1","versionstring":"19.0.1"})   QNetworkReply::NoError  Reply:  QNetworkReplyHttpImpl(0x28a6d504ff0)
2020-07-21 16:26:24:790 [ info nextcloud.sync.connectionvalidator ]:	** Application: ownCloud found:  QUrl("https://nextcloud.myserver.lan")  with version  "19.0.1" ( "19.0.1.1" )
2020-07-21 16:26:24:791 [ info nextcloud.sync.connectionvalidator ]:	QUrl("https://nextcloud.myserver.lan") has server version "19.0.1.1"
2020-07-21 16:26:24:793 [ info nextcloud.sync.account ]:	appPassword stored in keychain
2020-07-21 16:26:24:793 [ info nextcloud.sync.accessmanager ]:	6 "PROPFIND" "https://nextcloud.myserver.lan/remote.php/webdav/" has X-Request-ID "234fe68e-6d8f-41c4-a722-8ff9ebab6bf3"
2020-07-21 16:26:24:794 [ info nextcloud.sync.networkjob ]:	OCC::PropfindJob created for "https://nextcloud.myserver.lan" + "/" "OCC::ConnectionValidator"
2020-07-21 16:26:25:180 [ info nextcloud.sync.credentials.webflow ]:	request finished
2020-07-21 16:26:25:180 [ info nextcloud.sync.networkjob.propfind ]:	PROPFIND of QUrl("https://nextcloud.myserver.lan/remote.php/webdav/") FINISHED WITH STATUS "OK"
2020-07-21 16:26:25:181 [ info nextcloud.sync.accessmanager ]:	2 "" "https://nextcloud.myserver.lan/ocs/v1.php/cloud/capabilities?format=json" has X-Request-ID "b38670fb-c43d-44fc-b7ec-6146c553dbfe"
2020-07-21 16:26:25:181 [ info nextcloud.sync.networkjob ]:	OCC::JsonApiJob created for "https://nextcloud.myserver.lan" + "ocs/v1.php/cloud/capabilities" "OCC::ConnectionValidator"
2020-07-21 16:26:25:181 [ info nextcloud.sync.accessmanager ]:	2 "" "https://nextcloud.myserver.lan/ocs/v1.php/config?format=json" has X-Request-ID "36ae7f78-9cdc-4f03-a642-b98f4547829c"
2020-07-21 16:26:25:182 [ info nextcloud.sync.networkjob ]:	OCC::JsonApiJob created for "https://nextcloud.myserver.lan" + "ocs/v1.php/config" ""
2020-07-21 16:26:25:279 [ info nextcloud.sync.credentials.webflow ]:	request finished
2020-07-21 16:26:25:279 [ info nextcloud.sync.networkjob.jsonapi ]:	JsonApiJob of QUrl("https://nextcloud.myserver.lan/ocs/v1.php/cloud/capabilities?format=json") FINISHED WITH STATUS "OK"
2020-07-21 16:26:25:279 [ info nextcloud.sync.connectionvalidator ]:	Server capabilities QJsonObject({"bruteforce":{"delay":0},"core":{"pollinterval":60,"webdav-root":"remote.php/webdav"},"dav":{"chunking":"1.0"},"end-to-end-encryption":{"api-version":"1.1","enabled":true},"files":{"bigfilechunking":true,"blacklisted_files":[".htaccess"],"directEditing":{"etag":"cb10ae89879951cc75a571c216a42dd5","url":"https://nextcloud.myserver.lan/ocs/v2.php/apps/files/api/v1/directEditing"},"versioning":true},"files_sharing":{"api_enabled":true,"default_permissions":31,"federation":{"expire_date":{"enabled":true},"incoming":true,"outgoing":true},"group":{"enabled":true,"expire_date":{"enabled":true}},"group_sharing":true,"public":{"enabled":true,"expire_date":{"days":"7","enabled":true,"enforced":false},"expire_date_internal":{"enabled":false},"multiple_links":true,"password":{"askForOptionalPassword":false,"enforced":false},"send_mail":true,"upload":true,"upload_files_drop":true},"resharing":false,"sharebymail":{"enabled":true,"expire_date":{"enabled":true},"password":{"enabled":true},"upload_files_drop":{"enabled":true}},"sharee":{"query_lookup_default":false},"user":{"expire_date":{"enabled":true},"send_mail":false}},"notifications":{"admin-notifications":["ocs","cli"],"ocs-endpoints":["list","get","delete","delete-all","icons","rich-strings","action-web"],"push":["devices","object-data","delete"]},"ocm":{"apiVersion":"1.0-proposal1","enabled":true,"endPoint":"https://nextcloud.myserver.lan/ocm","resourceTypes":[{"name":"file","protocols":{"webdav":"/public.php/webdav/"},"shareTypes":["user","group"]}]},"password_policy":{"api":{"generate":"https://nextcloud.myserver.lan/ocs/v2.php/apps/password_policy/api/v1/generate","validate":"https://nextcloud.myserver.lan/ocs/v2.php/apps/password_policy/api/v1/validate"},"enforceNonCommonPassword":true,"enforceNumericCharacters":false,"enforceSpecialCharacters":false,"enforceUpperLowerCase":false,"minLength":8},"theming":{"background":"https://nextcloud.myserver.lan/core/img/background.png?v=0","background-default":true,"background-plain":false,"color":"#0082c9","color-element":"#0082c9","color-element-bright":"#0082c9","color-element-dark":"#0082c9","color-text":"#ffffff","favicon":"https://nextcloud.myserver.lan/core/img/logo/logo.svg?v=0","logo":"https://nextcloud.myserver.lan/core/img/logo/logo.svg?v=0","logoheader":"https://nextcloud.myserver.lan/core/img/logo/logo.svg?v=0","name":"Nextcloud","slogan":"ein sicherer Ort für all Deine Daten","url":"https://nextcloud.com"}})
2020-07-21 16:26:25:280 [ info nextcloud.sync.accessmanager ]:	2 "" "https://nextcloud.myserver.lan/ocs/v2.php/apps/files/api/v1/directEditing?format=json" has X-Request-ID "2a5adbdc-e4fe-4b7e-aba4-8264e8124900"
2020-07-21 16:26:25:282 [ info nextcloud.sync.networkjob ]:	OCC::JsonApiJob created for "https://nextcloud.myserver.lan" + "ocs/v2.php/apps/files/api/v1/directEditing" "OCC::Account"
2020-07-21 16:26:25:282 [ info nextcloud.sync.accessmanager ]:	2 "" "https://nextcloud.myserver.lan/ocs/v1.php/cloud/user?format=json" has X-Request-ID "a0090716-8e1d-4837-9c48-9468ddf3cfe8"
2020-07-21 16:26:25:282 [ info nextcloud.sync.networkjob ]:	OCC::JsonApiJob created for "https://nextcloud.myserver.lan" + "ocs/v1.php/cloud/user" "OCC::UserInfo"
2020-07-21 16:26:25:301 [ info nextcloud.sync.credentials.webflow ]:	request finished
2020-07-21 16:26:25:301 [ info nextcloud.sync.networkjob.jsonapi ]:	JsonApiJob of QUrl("https://nextcloud.myserver.lan/ocs/v1.php/config?format=json") FINISHED WITH STATUS "OK"
2020-07-21 16:26:25:301 [ info nextcloud.sync.connectionvalidator ]:	Determined user-visible host to be "nextcloud.myserver.lan"
2020-07-21 16:26:25:358 [ info nextcloud.sync.credentials.webflow ]:	request finished
2020-07-21 16:26:25:359 [ info nextcloud.sync.networkjob.jsonapi ]:	JsonApiJob of QUrl("https://nextcloud.myserver.lan/ocs/v2.php/apps/files/api/v1/directEditing?format=json") FINISHED WITH STATUS "OK"
2020-07-21 16:26:25:395 [ info nextcloud.sync.credentials.webflow ]:	request finished
2020-07-21 16:26:25:396 [ info nextcloud.sync.networkjob.jsonapi ]:	JsonApiJob of QUrl("https://nextcloud.myserver.lan/ocs/v1.php/cloud/user?format=json") FINISHED WITH STATUS "OK"
2020-07-21 16:26:25:396 [ info nextcloud.sync.accessmanager ]:	2 "" "https://nextcloud.myserver.lan/remote.php/dav/avatars/myaccount/128.png" has X-Request-ID "acb11f5b-0e23-40fc-ab18-c402f48b14b8"
2020-07-21 16:26:25:397 [ info nextcloud.sync.networkjob ]:	OCC::AvatarJob created for "https://nextcloudmyserver.lan" + "" "OCC::UserInfo"
2020-07-21 16:26:25:483 [ info nextcloud.sync.credentials.webflow ]:	request finished
2020-07-21 16:26:25:484 [ info nextcloud.sync.clientsideencryption ]:	Initializing
2020-07-21 16:26:25:484 [ info nextcloud.sync.server.capabilities ]:	E2EE API version: "1.1"
2020-07-21 16:26:25:486 [ info nextcloud.sync.clientsideencryption ]:	Retrieving public key from server
2020-07-21 16:26:25:487 [ info nextcloud.sync.accessmanager ]:	2 "" "https://nextcloud.myserver.lan/ocs/v2.php/apps/end_to_end_encryption/api/v1/public-key?format=json" has X-Request-ID "8d017ea0-e21e-4124-b30a-20471d27031b"
2020-07-21 16:26:25:487 [ info nextcloud.sync.networkjob ]:	OCC::JsonApiJob created for "https://nextcloud.myserver.lan" + "ocs/v2.php/apps/end_to_end_encryption/api/v1/public-key" "OCC::ClientSideEncryption"
2020-07-21 16:26:25:559 [ info nextcloud.sync.credentials.webflow ]:	request finished
2020-07-21 16:26:25:560 [ warning nextcloud.sync.networkjob ]:	QNetworkReply::ContentNotFoundError "Server hat \"404 Not Found\" auf \"GET https://nextcloud.myserver.lan/ocs/v2.php/apps/end_to_end_encryption/api/v1/public-key?format=json\" geantwortet" QVariant(int, 404)
2020-07-21 16:26:25:560 [ warning nextcloud.sync.credentials.webflow ]:	QNetworkReply::ContentNotFoundError
2020-07-21 16:26:25:560 [ warning nextcloud.sync.credentials.webflow ]:	"Error transferring https://nextcloud.myserver.lan/ocs/v2.php/apps/end_to_end_encryption/api/v1/public-key?format=json - server replied: Not Found"
2020-07-21 16:26:25:560 [ info nextcloud.sync.networkjob.jsonapi ]:	JsonApiJob of QUrl("https://nextcloud.myserver.lan/ocs/v2.php/apps/end_to_end_encryption/api/v1/public-key?format=json") FINISHED WITH STATUS "ContentNotFoundError Server hat \"404 Not Found\" auf \"GET https://nextcloud.myserver.lan/ocs/v2.php/apps/end_to_end_encryption/api/v1/public-key?format=json\" geantwortet"
2020-07-21 16:26:25:561 [ warning nextcloud.sync.networkjob.jsonapi ]:	Network error:  "ocs/v2.php/apps/end_to_end_encryption/api/v1/public-key" "Server hat \"404 Not Found\" auf \"GET https://nextcloud.myserver.lan/ocs/v2.php/apps/end_to_end_encryption/api/v1/public-key?format=json\" geantwortet" QVariant(int, 404)
2020-07-21 16:26:25:561 [ info nextcloud.sync.clientsideencryption ]:	No public key on the server
2020-07-21 16:26:25:561 [ info nextcloud.sync.clientsideencryption ]:	No public key, generating a pair.
2020-07-21 16:26:25:583 [ info nextcloud.sync.clientsideencryption ]:	Key correctly generated
2020-07-21 16:26:25:584 [ info nextcloud.sync.clientsideencryption ]:	Storing keys locally
2020-07-21 16:26:25:584 [ info nextcloud.sync.clientsideencryption ]:	Keys generated correctly, sending to server.
2020-07-21 16:26:25:584 [ info nextcloud.sync.clientsideencryption ]:	Getting the following array for the account Id "myaccount"
2020-07-21 16:26:25:585 [ info nextcloud.sync.clientsideencryption ]:	Returning the certificate
2020-07-21 16:26:25:585 [ info nextcloud.sync.clientsideencryption ]:	"-----BEGIN CERTIFICATE REQUEST-----SOMECSRCODE----END CERTIFICATE REQUEST-----\n"
2020-07-21 16:26:25:586 [ info nextcloud.sync.networkjob.sendcsr ]:	Sending the CSR "csr=-----BEGIN%20CERTIFICATE%20REQUEST-----SOMEKEYCODE-----END%20CERTIFICATE%20REQUEST-----%0A"
2020-07-21 16:26:25:587 [ info nextcloud.sync.accessmanager ]:	4 "" "https://nextcloud.myserver.lan/ocs/v2.php/apps/end_to_end_encryption/api/v1/public-key?format=json" has X-Request-ID "f7e0e864-f9b5-4ed2-9a4a-8e86c2a66c90"
2020-07-21 16:26:25:588 [ info nextcloud.sync.networkjob ]:	OCC::SignPublicKeyApiJob created for "https://nextcloud.myserver.lan" + "ocs/v2.php/apps/end_to_end_encryption/api/v1/public-key" "OCC::ClientSideEncryption"
2020-07-21 16:26:25:665 [ info nextcloud.sync.credentials.webflow ]:	request finished
2020-07-21 16:26:25:665 [ warning nextcloud.sync.networkjob ]:	QNetworkReply::ContentConflictError "Server hat \"409 Conflict\" auf \"POST https://nextcloud.myserver.lan/ocs/v2.php/apps/end_to_end_encryption/api/v1/public-key?format=json\" geantwortet" QVariant(int, 409)
2020-07-21 16:26:25:665 [ warning nextcloud.sync.credentials.webflow ]:	QNetworkReply::ContentConflictError
2020-07-21 16:26:25:666 [ warning nextcloud.sync.credentials.webflow ]:	"Error transferring https://nextcloud.myserver.lan/ocs/v2.php/apps/end_to_end_encryption/api/v1/public-key?format=json - server replied: Conflict"
2020-07-21 16:26:25:666 [ info nextcloud.sync.networkjob.storeprivatekey ]:	Sending CSR ended with "ocs/v2.php/apps/end_to_end_encryption/api/v1/public-key" "Server hat \"409 Conflict\" auf \"POST https://nextcloud.myserver.lan/ocs/v2.php/apps/end_to_end_encryption/api/v1/public-key?format=json\" geantwortet" QVariant(int, 409)
2020-07-21 16:26:25:666 [ info nextcloud.sync.clientsideencryption ]:	409
2020-07-21 16:26:27:607 [ info nextcloud.gui.updater ]:	Checking for available update
2020-07-21 16:26:27:607 [ info nextcloud.sync.accessmanager ]:	2 "" "https://updates.nextcloud.org/client/?version=2.7.0.20200716&platform=win32&oem=Nextcloud&versionsuffix=beta3-Win64&channel=beta&updatesegment=76" has X-Request-ID "0aa6f8e2-04ad-4426-83cb-011b9caca95c"
2020-07-21 16:26:27:804 [ info nextcloud.gui.updater ]:	Client is on latest version!
````

2. Web server error log:

3. Server logfile: nextcloud log (data/nextcloud.log):

**Comment by er-vin** [comment]
So from the log, this looks like something fails on the server when uploading the public key... This 409 is kind of surprising, you're sure there's no such key on the server?

@georgehrke any idea why the server would reply a 409 after a POST on `/ocs/v2.php/apps/end_to_end_encryption/api/v1/public-key?format=json`?

**Comment by er-vin** [comment]
Ah! One thing for you @spackmat 

Could you please try again with debug enabled for the logs? I somehow doubt it'll bring up much more in that case but you never know.
Please also try to locate in the server logs when that 409 is sent back, we might have more clue as to why the server decide to go with it.

**Comment by georgehrke** [comment]
> @georgehrke any idea why the server would reply a 409 after a POST on `/ocs/v2.php/apps/end_to_end_encryption/api/v1/public-key?format=json`?

That route throws a 409 when a public key already exists:
https://github.com/nextcloud/end_to_end_encryption/blob/2b19530bb3b70d7e5fb05d6638de86bec5feea91/lib/Controller/KeyController.php#L219

@spackmat Can you connect to the server, navigate to Nextcloud's root folder, execute the following command please:
`php occ files:scan-app-data` and then try again?

**Comment by spackmat** [comment]
@georgehrke yes, that command fixed the problem! Thanks.

I deleted the three folders `./meta-data`, `./public-keys`, `./private-keys` inside `appdata_{instanceId}/end_to_end_encryption` and then deactivated and removed the whole old E2EE app. After that also the `appdata_{instanceId}/end_to_end_encryption` folder was deleted. Maybe that was the problem here?

**Comment by er-vin** [comment]
> > @georgehrke any idea why the server would reply a 409 after a POST on `/ocs/v2.php/apps/end_to_end_encryption/api/v1/public-key?format=json`?
> 
> That route throws a 409 when a public key already exists:
> https://github.com/nextcloud/end_to_end_encryption/blob/2b19530bb3b70d7e5fb05d6638de86bec5feea91/lib/Controller/KeyController.php#L219
> 
> @spackmat Can you connect to the server, navigate to Nextcloud's root folder, execute the following command please:
> `php occ files:scan-app-data` and then try again?

Maybe we want to update the "cleanup procedure" in the 1.5.2-beta1 release notes and the forum post then? I think that step was missing.

**Comment by spackmat** [comment]
And directly to the next problem: I encrypted a test folder, uploaded a pdf file while the app is running and everything works. Then I closed the app, copied a new file to that encrypted folder and removed the other one. Then I started the client again and got a 403 for the DELETE request. While I was writing this message, the client tried again and succeeded. Seems to be still a bit quirky.

But the good part is: This is the first time, I got the E2EE to work with the desktop client after years and hours of testing attempts. So hooray and thanks for the progress.
